http-gateways/IPLD_GATEWAY.md

```diff
@@ -0,0 +1,196 @@
+- `application/vnd.ipld.dag-json`: Return the block specified by the path encoded in `dag-json`.
+- `application/vnd.ipld.dag-cbor`: Return the block specified by the path ecnoded in `dag-cbor`.
+
+### Request Query Parameters
+
+#### `format` (request query parameter)
+
+Optional, `format=<format>` can be used to request specific encodings.
+
+This is a URL-friendly alternative to sending `Accept: application/vnd.ipld.<format>` header, see [Accept](#accept-request-header) for more details.
+
+### Path Segments and Path Segment Parameters
+
+Path segments are used for IPLD data model [traversal](https://ipld.io/docs/data-model/traversal/).
+Each segment is separated by a `/` and contains a utf8 `name` followed by an optional set of parameters using the [Matrix URI format](https://www.w3.org/DesignIssues/MatrixURIs.html).
+
+Example:
+
+```
+/ipld/bafywhatever/foo/bar;extra=thing;whatever=here/
+```
+
+Path segments can have additional parameters added to them by separating them using semicolons (`;`) and having key-value pairs separated by an `=`.
+This format is based on the [Matrix URI proposal from the w3c](https://www.w3.org/DesignIssues/MatrixURIs.html).
+Note that these parameters are stripped from the segment name when passed to any underlying traversal code.
+
+This spec only perscribes two reserved parameter names: `ADL` for specifying [Advanced Data Layouts](https://ipld.io/docs/advanced-data-layouts/) to process the data with, and `schema` to specify an [IPLD schema](https://ipld.io/docs/schemas/intro/) to use to interpret the data.
+Other names may be specified in future specs that build upon this one.
+
+#### ADL (segment parameter)
+
+Segments may contain an `adl` key which points to a name of an [Advanced Data Layout](https://ipld.io/docs/advanced-data-layouts/intro/) to process the node with.
+
+The supported ADL names will vary based on gateway.
+
+An example value would be `;adl=hamt` to specify the [HAMT](https://ipld.io/specs/advanced-data-layouts/hamt/) ADL that's used to represent large maps.
+
+For example `/ipld/bafyreic672jz6huur4c2yekd3uycswe2xfqhjlmtmm5dorb6yoytgflova;adl=hamt/yes` (taken from the HAMT examples), should resolve to the following:
+
+```
+[
+  {
+    "line": 9,
+    "column": 501
+  }
+]
+```
+
+#### Schema (segment parameter)
+
+Segments may contain a `schema` key which points to the `CID` of an [IPLD Schema](https://ipld.io/docs/schemas/) in its [DMT](https://ipld.io/specs/schemas/#dsl-vs-dmt) form.
+If a `schema` key is provided, there must also be a `type` parameter which references one of the named Types within the IPLD Schema.
+
+The node at that point in the traversal will then be transformed by the schema, and any typed nodes it links to will also be transformed by their respective schemas.
+
+For example, given the following schema (note it is written in DSL form, but must be converted to the DMT in order to be refernced):
+
+```ipldschema
+type Example struct {
```

Comment:
nit: I've read this section and tbh have no idea what is the value to end user – CBOR traversal and field resolution with extra steps so the output looks a certain way? 

In ADL section we have good use case "ADL that's used to represent large maps" – we need similar real world example for schemas.

What would a schema be useful for irl?  I feel the spec here needs better `Example`, so the value is obvious.

Reply:
@rvagg @warpfork would you be able to comment on real world uses of IPLD Schema that would be relevant here?

Reply:
One use case for schemas is to use the `representation` functionality to render data in more human/application readable formats from formats that are more size efficient.

e.g. some things might be using a `listpairs` representation which would look like an array of arrays by default. But with a schema you can transform the representation to be more human readable.

I'm gonna be doing stuff along this line for the Prolly Tree work where we'll be encoding tree nodes more efficiently, but having a way to put them through a schema before the application code starts working with them.